packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -219,282 +209,70 @@ export class OdspDocumentService implements IDocumentService {
-            // cache entry to be treated as expired after 1 hour.
-            _response.joinSessionResponse.refreshSessionDurationSeconds =
-                _response.joinSessionResponse.refreshSessionDurationSeconds ?? 3600;
-            return {
-                ..._response,
-                refreshAfterDeltaMs: this.calculateJoinSessionRefreshDelta(
-                    _response.entryTime, _response.joinSessionResponse.refreshSessionDurationSeconds),
-            };
-        };
-        let response = await getResponseAndRefreshAfterDeltaMs();
-        // This means that the cached entry has expired(This should not be possible if the response is fetched
-        // from the network call). In this case we remove the cached entry and fetch the new response.
-        if (response.refreshAfterDeltaMs <= 0) {
-            this.cache.sessionJoinCache.remove(this.joinSessionKey);
-            response = await getResponseAndRefreshAfterDeltaMs();
-        }
-        if (!disableJoinSessionRefresh) {
-            const props = {
-                entryTime: response.entryTime,
-                refreshSessionDurationSeconds:
-                    response.joinSessionResponse.refreshSessionDurationSeconds,
-                refreshAfterDeltaMs: response.refreshAfterDeltaMs,
-            };
-            if (response.refreshAfterDeltaMs > 0) {
-                this.scheduleJoinSessionRefresh(response.refreshAfterDeltaMs)
-                    .catch((error) => {
-                        const canRetry = canRetryOnError(error);
-                        // Only record error event in case it is non retriable.
-                        if (!canRetry) {
-                            this.mc.logger.sendErrorEvent({
-                                eventName: "JoinSessionRefreshError",
-                                details: JSON.stringify(props),
-                            },
-                            error,
-                            );
-                        }
-                    });
-            } else {
-                // Logging just for informational purposes to help with debugging as this is a new feature.
-                this.mc.logger.sendTelemetryEvent({
-                    eventName: "JoinSessionRefreshNotScheduled",
-                    details: JSON.stringify(props),
-                });
-            }
+        if (this.socketModuleP === undefined) {
+            this.socketModuleP = new Deferred();
+        } else {
+            await this.socketModuleP.promise;
         }
-        return response.joinSessionResponse;
-    }
-
-    private calculateJoinSessionRefreshDelta(responseFetchTime: number, refreshSessionDurationSeconds: number) {
-        // 30 seconds is buffer time to refresh the session.
-        return responseFetchTime + ((refreshSessionDurationSeconds * 1000) - 30000) - Date.now();
+        const deltaStream = await this.getDelayLoadedDeltaStream();
```

Comment:
If anything throws in getDelayLoadedDeltaStream(), then await this.socketModuleP would hang forever.
I'd suggest either using LazyPromise() construct here, or (I believe we confirmed earlier that doing concurrent loading is fine) allow concurrent loading of module

Reply:
Actually, you can use regular promises here. Something like that:

```ts
socketModuleP: Promise<something>; // return type of getDelayLoadedDeltaStream()

if (this.socketModuleP == undefined) {
    this.socketModuleP = this.getDelayLoadedDeltaStream();
}

return this.socketModuleP;
```